### ISSUE TITLE
Import non-weight values of Normalisation neuron

### ISSUE THREAD

**Issue by jzstark** [owlbarn/owl#288]
In neural network, a Normalisation neuron contains `mu` and `var` parameters, besides `gamma` and `beta`. Its behaviour varies in different mode: it normalises input with average and variance of current batch in training mode and keep updating the moving `mu` and `var`;  in inference mode it normalises input directly with the calculated `mu` and `var` from training phase. Unlike `gamma` and `beta`, these two are not weight values.

One implication of this difference is that, when we are importing values from existing pre-trained models, we cannot directly use `Neuron.update` to import the value of `mu` and `var` into a network. My current solution is to manually update these values in the normalisation neurons in a network, and then save the network itself instead of its weight.  As a result, for image classification networks such as Inception and ResNet, by saving the network, we cannot then change its input sizes. 

So I'm looking at either a better alternative to this current practice in importing values from existing models (e.g. maybe using the OpenVX/NNEF format standards Liang has mentioned), or a mechanism to import non-weight trainable values following that of existing frameworks such as Keras/MxNet/etc. 

In Keras, each weight variable is assigned "trainable" or "non-trainable" -- the latter only consists of `mean` and `variance` in the BatchNormalization layer.  The weight loading/saving does not distinguish between these two kinds of weights.

**Comment by pvdhove** [comment]
Here is what I could find in other frameworks:
- in Keras, each layer has two list attributes `trainable_weights` and `non_trainable_weights`. The function `get_weights` simply returns the concatenation of the lists. Weights are wrapped in a `Variable` (similar to a `Var` node from Owl's CG) and must be added to a layer with a function `add_weights(shape, name, initializer, regularizer, trainable)`. This function returns the `Variable` object and adds it to either `trainable_weights` or `non_trainable_weights`. Since the value of the weights is wrapped into a `Variable`, both the weight and its reference in the list can point to the same ndarray value (in Owl, the weight is the ndarray value so that is not currently possible).
- in Tensorflow, weights are `variable`s and their values are saved when you save a model. The training part is a bit uglier.
- PyTorch distinguishes parameters (W, b) from 'buffers' (mean, var). Buffers are only explicitly used for the normalisation neuron. The `save` function saves both parameters and buffers. BatchNorm also overwrites the `load_from_state_dict` function and is the only neuron doing so.

I think it is good to keep Algodiff distinct from CG so that we can choose if we want to use the optimisations or not. Wrapping each weight in some kind of struct `weight {value:t; trainable:bool; }` is possible but would make the code much more verbose (we would need to call `l.w.value` instead of `l.w` everywhere and the initialisation would be much uglier).

1) One solution, like PyTorch, is to have a `save`/`load` function for each neuron which is only different in `Normalisation`.
2) Another one would be to add a `non_trainable_weights` array (which would be `[||]` by default) next to `mkpar` and add a `get_weights` function concatenating both. The `update` function would then need a boolean to express if it should update non-trainable weights as well.
3) Similarly, we could have a `weights: t array` and a `trainable: bool array` attribute for each neuron, and then `mkpar`, `mkadj`, `mrpri`, `non_trainable_weights` can all be expressed in terms of `filter` and `map` of `weights` and `trainable`.

My opinion is that 2) looks clean and is the closest one to what other frameworks do, let me know if that sounds OK to you :)

**Comment by ryanrhymes** [comment]
@pvdhove Cool, thanks for investigating the details of these frameworks. I think all three solutions should work. 2) is good, I think 1) also looks very neat. What do you think are the pros and cons comparing 1) and 2) ... i am quite neutral.

**Comment by pvdhove** [comment]
Solution 1) is the one requiring the least amount of changes but is slightly more hacky (it is just wrapping an extra function to make `Normalisation` work). Solution 2) would be clean if a lot of neurons had non trainable weights, but maybe adding an empty array `non_trainable_weights` to each neuron only because of `Normalisation` is a bit too much. I'll implement 1) and you can judge afterwards :)

**Comment by ryanrhymes** [comment]
OK, I see. Yes, I was simply considering w.r.t to the amount of code you need to write. I think the decision depends on how common the non-trainable weights are used in neurons. There are still some neurons missing certainly but owl already implements quite a lot, the only neuron atm uses non-trainable weights is normalisation neuron. Do you know are there any other special neurons will utilise non-trainable weight?

@jzstark what's your knowledge and opinion on this?

**Comment by jzstark** [comment]
I searched all the neurons in Keras and so far only the Normalisation neuron has non-trainable weights so 1) is looks a quite nice solution to me. But I'm also thinking about "freezing" a model in training. I'm not quite clear about how it is implemented, but one possible way could be to make the certain weights non-trainable in the other neurons.

**Comment by ryanrhymes** [comment]
@jzstark good point, if we take freezing function into account, maybe 3) is actually better? Can you two have a meeting tomorrow to see if these two functionality can be sorted out together actually?

**Comment by pvdhove** [comment]
In PyTorch, only BatchNorm and InstanceNorm (which is really similar) use the 'buffer' (=non-trainable weights) attribute.
In Keras, every weight is trainable or non-trainable but every layer also has a `trainable` boolean. I believe the `trainable` attribute of **weights** is not supposed to be changed, I don't think it makes much sense to freeze only some weights in a layer (I will check though), that attribute is only used by the optimisation engine. The `trainable` attribute of **layers** can easily be changed dynamically though. I will check how that is done in PyTorch.

**Comment by pvdhove** [comment]
In other frameworks, the freezing of weights is not related to the freezing of layers. Even if we had a boolean array `trainable` to express which weights are trainable, it would be annoying to set all its values to `false` to make a layer non-trainable because we would need to remember what is the initial trainability of each weight to be able to set the layer trainable again. It makes more sense to have a separate boolean value `trainable` for the whole **layer**, and then just ignore a layer at training time if `trainable = false` (which is what is done in Keras).
For the normalisation neuron, solution 2) is clean to save the weights but it does not solve the loading of the weights in a clean way and we have to use solution 1) for it anyway (because we don't want to add a boolean attribute to `update`). So we'll just implement solution 1) :)

**Comment by ryanrhymes** [comment]
OK, let's go for solution (1) which seems the most effective and clean way to address this.

**Comment by ryanrhymes** [comment]
@jzstark since our understanding on these mechanisms have been improved significantly thanks to Pierre's recent research on other frameworks. Maybe it is time to fix this issue with pierre together beside the aeos stuff? https://github.com/owlbarn/owl/issues/239

**Comment by jzstark** [comment]
Yes indeed, we discussed about it and agree that we should add the model freeze function now to see if the weight-saving improvement Pierre has made requires any further modification.
